routers/api/v1/repo/commits.go

```diff
@@ -48,13 +48,61 @@ func GetSingleCommit(ctx *context.APIContext) {
 	//   "404":
 	//     "$ref": "#/responses/notFound"
 
+	sha := ctx.Params(":sha")
+	if len(sha) == 0 {
+		ctx.Error(http.StatusBadRequest, "ref not given", nil)
+		return
+	}
+	getCommit(ctx, sha)
+}
+
+// GetSingleCommitByRef get a commit via ref
+func GetSingleCommitByRef(ctx *context.APIContext) {
+	// swagger:operation GET /repos/{owner}/{repo}/commits/{ref} repository repoGetSingleCommitByRef
+	// ---
+	// summary: Get a single commit from a repository
+	// produces:
+	// - application/json
+	// parameters:
+	// - name: owner
+	//   in: path
+	//   description: owner of the repo
+	//   type: string
+	//   required: true
+	// - name: repo
+	//   in: path
+	//   description: name of the repo
+	//   type: string
+	//   required: true
+	// - name: ref
+	//   in: path
+	//   description: a git ref
+	//   type: string
+	//   required: true
+	// responses:
+	//   "200":
+	//     "$ref": "#/responses/Commit"
+	//   "400":
+	//     "$ref": "#/responses/error"
+	//   "404":
+	//     "$ref": "#/responses/notFound"
+
+	ref := ctx.Params("ref")
+	if len(ref) == 0 {
```

Comment:
Maybe we should validate `ref` value as well? I did some tests and so far the only invalid strings I could generate were those:

- starting or ending with `/`
- containing spaces, `:`, `*`, `[`, `~`, `\` or `?`.
- containing `"` (it's accepted, but the command fails).
- containing parts starting or ending with `.` (e.g. `refs/branches/.my/branch`).

Also, are we accepting any kind of refs, even if they are ambiguous? (e.g. branch and tag with the same name).

Reply:
we should think of souch things when **PUSH** ref's but it ref do not exist it simply get a 404 ...

this is only a get func

Reply:
But you could still probably do command injection attack via this probably if anything is accepted and passed to command line

Reply:
then it already could be done **NOW**

Reply:
ref: `trash; git execute this!!`

Reply:
not so easy, we probably do some sanity tests for command line arguents but anyway we should allow only sha here not branch names, tags etc

Reply:
@lafriks `GetSingleCommitBySHA` should be restricted to sha then and `GetSingleCommitByRef` has to check if its a valid rev ...

Reply:
What is a valid git commit reference is really rather difficult to ascertain. If you only want SHA then that's fine however, if you want to allow branches - then... the only answer is to pass the string to `git rev-parse --verify`. This seems mostly safe - remotes are available to be queried however.

Reply:
@6543 

> ref: `trash; git execute this!!`

It doesn't work like that the - the `trash; git execute this!!` is passed to git as a single argument - equivalent to `git rev-parse --verify "trash; git execute this!!"`

Reply:
https://github.com/go-gitea/gitea/blob/master/modules/validation/binding.go has git ref checking if needed

Reply:
this would need a litle refactor so this func could be used outside macaron